File: src/dart/matmul.dart
import 'dart:typed_data';

List<Float64List> matgen(int n) {
	var a = List<Float64List>.generate(n, (i)=>Float64List(n), growable: false);
	final tmp = 1.0 / n / n;
	for (var i = 0; i < n; ++i) {
		for (var j = 0; j < n; ++j) {
			a[i][j] = tmp * (i - j) * (i + j);
		}
	}
	return a;
}

List<Float64List> matmul(a, b) {
	final n = a.length, p = a[0].length, m = b[0].length;
	var c = List<Float64List>.generate(n, (i)=>Float64List(m), growable: false);
	for (var i = 0; i < n; ++i) {
		var ci = c[i];
		for (var k = 0; k < p; ++k) {
			final t = a[i][k], bk = b[k];
			for (var j = 0; j < m; ++j)
				ci[j] += t * bk[j];
		}
	}
	return c;
}

main() {
	var n = 1500;
	final a = matgen(n);
	final b = matgen(n);
	final c = matmul(a, b);
	print(c[n>>1][n>>1]);
}

File: src/zig/matmul.zig
const std = @import("std");

fn mat_alloc(n: u32) [][]f64 {
	var a: [][]f64 = undefined;
	a = std.heap.page_allocator.alloc([]f64, n) catch unreachable;
	for (a) |*row| {
		row.* = std.heap.page_allocator.alloc(f64, n) catch unreachable;
	}
	return a;
}

fn matgen(n: u32) [][]f64 {
	var a = mat_alloc(n);
	const f: f64 = @floatFromInt(n);
	const tmp  = 1.0 / f / f;
	for (0..n) |i| {
		const ii: i64 = @bitCast(i);
		for (0..n) |j| {
			const jj: i64 = @bitCast(j);
			a[i][j] = tmp * @as(f64, @floatFromInt(ii - jj)) * @as(f64, @floatFromInt(ii + jj));
		}
	}
	return a;
}

fn matmul(n: u32, a: [][]f64, b: [][]f64) [][]f64 {
	var c = mat_alloc(n);
	for (0..n) |i| {
		for (0..n) |j| {
			c[i][j] = 0.0;
		}
	}
	for (0..n) |i| {
		var ci = c[i];
		for (0..n) |k| {
			const aik = a[i][k];
			const bk = b[k];
			for (0..n) |j| {
				ci[j] += aik * bk[j];
			}
		}
	}
	return c;
}

pub fn main() !void {
	const n = 1500;
	const a = matgen(n);
	const b = matgen(n);
	const c = matmul(n, a, b);
	std.debug.print("{}\n", .{c[n>>1][n>>1]});
	// FIXME: deallocate!
}
